starboard/shared/libiamf/iamf_decoder_utils.cc

```diff
@@ -0,0 +1,638 @@
+    int bytes_read = ReadLeb128Internal(
+        buf_ + pos_, ptr,
+        std::min(static_cast<uint64_t>(RemainingSize()), sizeof(uint32_t)));
+    if (bytes_read < 0) {
+      return false;
+    }
+    pos_ += bytes_read;
+    return true;
+  }
+
+  bool ReadString(std::string* str) {
+    if (!HasBytes(1) || !str) {
+      return false;
+    }
+
+    // The size of the string is capped to 128 bytes.
+    // https://aomediacodec.github.io/iamf/v1.0.0-errata.html#convention-data-types
+    const int kMaxIamfStringSize = 128;
+    int bytes_read = ReadStringInternal(
+        buf_ + pos_, str, std::min(RemainingSize(), kMaxIamfStringSize));
+    if (bytes_read < 0) {
+      return false;
+    }
+    pos_ += bytes_read;
+    return true;
+  }
+
+  bool SkipBytes(size_t size) {
+    if (!HasBytes(size)) {
+      return false;
+    }
+    pos_ += size;
+    return true;
+  }
+
+  bool SkipLeb128() {
+    uint32_t val;
+    return ReadLeb128(&val);
+  }
+
+  bool SkipString() {
+    std::string str;
+    return ReadString(&str);
+  }
+
+  size_t size() const { return size_; }
+  size_t pos() const { return pos_; }
+  const uint8_t* buf() const { return buf_; }
+
+ private:
+  bool HasBytes(size_t size) const { return size + pos_ <= size_; }
+  int RemainingSize() const { return size_ - pos_; }
+  inline uint32_t ByteSwap(uint32_t x) const {
+#if defined(COMPILER_MSVC)
+    return _byteswap_ulong(x);
+#else
+    return __builtin_bswap32(x);
+#endif
+  }
+
+  // Decodes an Leb128 value and stores it in |value|. Returns the number of
+  // bytes read, capped to |max_bytes_to_read|. Returns the number of bytes
+  // read, or -1 on error.
+  static int ReadLeb128Internal(const uint8_t* buf,
+                                uint32_t* value,
+                                const int max_bytes_to_read) {
+    SB_DCHECK(buf);
+    SB_DCHECK(value);
+
+    *value = 0;
+    bool error = true;
+    size_t i = 0;
+    for (; i < max_bytes_to_read; ++i) {
+      uint8_t byte = buf[i];
+      *value |= ((byte & 0x7f) << (i * 7));
+      if (!(byte & 0x80)) {
+        error = false;
+        break;
+      }
+    }
+
+    if (error) {
+      return -1;
+    }
+    return i + 1;
+  }
+
+  // Reads a c-string into |str|. Returns the number of bytes read, capped to
+  // 128 bytes, or -1 on error.
+  static int ReadStringInternal(const uint8_t* buf,
+                                std::string* str,
+                                int max_bytes_to_read) {
+    SB_DCHECK(buf);
+    SB_DCHECK(str);
+
+    str->clear();
+    str->resize(max_bytes_to_read);
+
+    int bytes_read = ::starboard::strlcpy(
+        str->data(), reinterpret_cast<const char*>(buf), max_bytes_to_read);
+    if (bytes_read == max_bytes_to_read) {
+      // Ensure that the string is null terminated.
+      if (buf[bytes_read] != '\0') {
+        return false;
+      }
+    }
+    str->resize(bytes_read);
+
+    // Account for null terminator.
+    return ++bytes_read;
+  }
+
+  int pos_ = 0;
```

Comment:
nit: move pos_ below the two variables as they will be initialized in the ctor and never modified after.